runtime/core/src/main/java/io/atlasmap/core/AtlasPath.java

```diff
@@ -230,44 +317,170 @@ public String getCollectionSegment() {
-                    getSegments().set(i, cleanPathSegment(segment) + JAVAPATH_ARRAY_START + index + JAVAPATH_ARRAY_END);
+                String part = cleanPathSegment(getSegments().get(i));
+                if (part != null && part.equals(cleanPathSegment(segment))) {
+                    getSegments().set(i, cleanPathSegment(segment) + PATH_ARRAY_START + index + PATH_ARRAY_END);
                 }
             }
-        } else if (segment.contains(JAVAPATH_LIST_START) && segment.contains(JAVAPATH_LIST_END)) {
+        } else if (segment.contains(PATH_LIST_START) && segment.contains(PATH_LIST_END)) {
             for (int i = 0; i < getSegments().size(); i++) {
-                String part = getSegments().get(i);
-                if (cleanPathSegment(part).equals(cleanPathSegment(segment))) {
-                    getSegments().set(i, cleanPathSegment(segment) + JAVAPATH_LIST_START + index + JAVAPATH_LIST_END);
+                String part = cleanPathSegment(getSegments().get(i));
+                if (part != null && part.equals(cleanPathSegment(segment))) {
+                    getSegments().set(i, cleanPathSegment(segment) + PATH_LIST_START + index + PATH_LIST_END);
                 }
             }
         } else {
-            throw new IllegalArgumentException("JavaPath segment is not a List or Array segment");
+            throw new IllegalArgumentException("PathUtil segment is not a List or Array segment");
         }
     }
 
-    @Override
     public String toString() {
-        StringBuffer buffer = new StringBuffer();
+        StringBuilder builder = new StringBuilder();
 
         int i = 0;
+
+        if (getSegments().size() > 1) {
+            builder.append(PATH_SEPARATOR);
+        }
+
         for (String part : getSegments()) {
-            buffer.append(part);
+            builder.append(part);
             if (i < (getSegments().size() - 1)) {
-                buffer.append(JAVAPATH_SEPARATOR);
+                builder.append(PATH_SEPARATOR);
             }
             i++;
         }
-        return buffer.toString();
+        return builder.toString();
+    }
+
+    public String getOriginalPath() {
+        return originalPath;
+    }
+
+    public static boolean isArraySegment(String segment) {
+        return isCollectionSegment(segment) && segment.contains(PATH_ARRAY_START);
+    }
+
+    public static boolean isListSegment(String segment) {
+        return isCollectionSegment(segment) && segment.contains(PATH_LIST_START);
+    }
+
+    public static boolean isMapSegment(String segment) {
+        return isCollectionSegment(segment) && segment.contains(PATH_MAP_START);
+    }
+
+    public static String findNextSegment(String segment, String path) {
+        AtlasPath pathUtil = new AtlasPath(path);
+        String parentSegment = null;
+        for (String currentSegment : pathUtil.segments) {
+            if (parentSegment == null || !segment.equals(parentSegment)) {
```

Comment:
Yeah I think Codacy is right here, this loop looks super weird. @sunilmeddr I know it was just copied, but can you turn over the condition and just return in 'if'?
https://www.codacy.com/app/atlasmapio/atlasmap/pullRequest?prid=1131462